File: src/layers/OscLayer.py
import numpy as np
import tensorflow as tf

class OscLayer(object):
    def __init__(self, num_osc, N, dt):
        self.N = N
        self.dt = dt
        #print(dt)
        self.num_osc = num_osc

    def get_signal(self, freq):
        Z = np.zeros((self.num_osc, self.N), dtype = np.complex128)
        #Zi = np.zeros((self.num_osc, self.N))
        r = np.ones((self.num_osc, self.N))
        phi = np.zeros((self.num_osc, self.N))
        Z[:, 0] = r[:, 0]*np.exp(1j*phi[:, 0])
        #Zi[:, 0] = r[:, 0]*np.exp(1j*phi[:, 0]).imag
        for i in range(self.N-1):
            r[:, i+1] = r[:, i] + (1-r[:, i]**2)*r[:, i]*self.dt
            #print(r[:, i+1])
            #print("\n")
            phi[:, i+1] = phi[:, i] + freq*self.dt
            #print(phi[:, i+1])
            #print("\t")
            Z[:, i+1] = r[:, i+1]*np.exp(1j*phi[:, i+1])
            #Zi[:, i+1] = r[:, i+1]*np.exp(1j*phi[:, i+1]).imag
        return Z#r* 1j*Zi

File: src/cpp_model/OscLayer.cpp
#include "OscLayer.h"

OscLayer::OscLayer(int n_o, int n, float dT){
    num_osc = n_o;
    dt = dT;
    N = n;
    r = new float[num_osc];
    phi = new float[num_osc];
    Z = new float*[num_osc];
    for(int i=0; i<num_osc; i++){
        Z[i] = new float[N];
        r[i] = 1.0;
        phi[i] = 0.0;
    }
}

std::complex** OscLayer::forwardPropagation(float *omega){
    std::complex<float> iota(0, 1); 
    float power = 2.0;
    for(int i=0; i<N; i++){
        for(int j =0; j<num_osc; j++){
            r[j] = r[j] + (1-pow(abs(r[j]), power))*r[j]*dt;
            phi[j] = phi[j] + freq*dt;
            Z[j][i] = r[i]*std::exp(phi[i]*iota);
        }  
    } 
    return Z;
}
